### ISSUE TITLE
Combining `join` and indexing generates far too many records

### ISSUE THREAD

**Issue by FRidh** [ContinuumIO/blaze#975]
I join two tables. Before taking the join I select only two columns from one of the tables. Instead of getting 2433 records (entry 10) I get 19524825 (entry 11).

```py

In [1]: from blaze import Data, join

In [2]: d = Data("sqlite:///sonair.sqlite3")

In [3]: d.event.head()
Out[3]: 
          name operation  datetime_nearest_receiver aircraft  temperature_2m  \
0  24_002_A320       S16 2014-05-15 07:46:15.100000     A320             7.3   
1  24_005_A321       S16 2014-05-15 08:08:23.200000     A321             7.5   
2  24_006_A321       S16 2014-05-15 08:11:50.250000     A321             7.5   
3  24_007_A333       S16 2014-05-15 08:15:52.850000     A333             8.0   
4  24_008_F100       S16 2014-05-15 08:21:51.650000     F100             8.0   
5  24_011_B762       S16 2014-05-15 08:55:40.500000     B762             9.0   
6  24_012_A320       S16 2014-05-15 09:22:46.050000     A320             9.5   
7  24_016_A320       S16 2014-05-15 10:27:42.100000     A320            10.0   
8  24_017_B764       S16 2014-05-15 10:35:20.050000     B764            10.1   
9  24_020_A333       S16 2014-05-15 11:03:01.650000     A333             9.8   

   temperature_5cm  humidity  windspeed  pressure  include  
0              8.5      91.7        1.6     978.9        1  
1              9.4      84.8        1.2     978.9        1  
2              9.4      84.8        2.3     978.9        1  
3             11.1      84.7        2.3     979.0        1  
4             11.1      84.7        1.8     979.0        1  
5             14.9      77.6        2.2     978.7        1  
6             14.1      73.5        8.3     978.7        1  
7             16.8      72.9        2.4     978.5        1  
8             13.6      75.9        2.8     978.5        1  
9             11.3      68.4        1.8     978.4        1  

In [4]: d.event.count()
Out[4]: 2675

In [5]: d.operation.head()
Out[5]: 
  name  runway takeoff datetime_nearest_close
0  S28      28    True                      A
1  S16      16    True                      Q
2  L14      14   False                      I

In [6]: d.operation.count()
Out[6]: 3

In [7]: d.event[d.event.include==True]
Out[7]: 
           name operation  datetime_nearest_receiver aircraft  temperature_2m  \
0   24_002_A320       S16 2014-05-15 07:46:15.100000     A320             7.3   
1   24_005_A321       S16 2014-05-15 08:08:23.200000     A321             7.5   
2   24_006_A321       S16 2014-05-15 08:11:50.250000     A321             7.5   
3   24_007_A333       S16 2014-05-15 08:15:52.850000     A333             8.0   
4   24_008_F100       S16 2014-05-15 08:21:51.650000     F100             8.0   
5   24_011_B762       S16 2014-05-15 08:55:40.500000     B762             9.0   
6   24_012_A320       S16 2014-05-15 09:22:46.050000     A320             9.5   
7   24_016_A320       S16 2014-05-15 10:27:42.100000     A320            10.0   
8   24_017_B764       S16 2014-05-15 10:35:20.050000     B764            10.1   
9   24_020_A333       S16 2014-05-15 11:03:01.650000     A333             9.8   
10  24_023_A333       S16 2014-05-15 11:21:32.650000     A333             9.7   

    temperature_5cm  humidity  windspeed  pressure  include  
0               8.5      91.7        1.6     978.9        1  
1               9.4      84.8        1.2     978.9        1  
2               9.4      84.8        2.3     978.9        1  
3              11.1      84.7        2.3     979.0        1  
4              11.1      84.7        1.8     979.0        1  
5              14.9      77.6        2.2     978.7        1  
6              14.1      73.5        8.3     978.7        1  
7              16.8      72.9        2.4     978.5        1  
8              13.6      75.9        2.8     978.5        1  
9              11.3      68.4        1.8     978.4        1  
...

In [8]: d.event[d.event.include==True].count()
Out[8]: 2433

In [9]: join( d.event, d.operation, 'operation', 'name').count()
Out[9]: 2675

In [10]: join( d.event[d.event.include==True], d.operation, 'operation', 'name').count()
Out[10]: 2433

In [11]: join( d.event[d.event.include==True], d.operation[['name', 'datetime_nearest_close']], 'operation', 'name').count()
Out[11]: 19524825

In [12]: join( d.event, d.operation[['name', 'datetime_nearest_close']], 'operation', 'name').count()
Out[12]: 2675

In [13]: join( d.operation[['name', 'datetime_nearest_close']], d.event[d.event.include==True], 'name', 'operation').count()
/home/rifr/anaconda3/envs/sonair/lib/python3.4/site-packages/sqlalchemy/sql/base.py:508: SAWarning: Column 'name' on table <sqlalchemy.sql.selectable.Select at 0x7ff5d3104898; Select object> being replaced by Column('name', TEXT(), table=<Select object>), which has the same key.  Consider use_labels for select() statements.
  self[column.key] = column
/home/rifr/anaconda3/envs/sonair/lib/python3.4/site-packages/sqlalchemy/sql/base.py:508: SAWarning: Column 'name' on table <sqlalchemy.sql.selectable.Alias at 0x7ff5d3104630; alias_6> being replaced by Column('name', TEXT(), table=<alias_6>), which has the same key.  Consider use_labels for select() statements.
  self[column.key] = column
/home/rifr/anaconda3/envs/sonair/lib/python3.4/site-packages/sqlalchemy/sql/base.py:508: SAWarning: Column 'name' on table <sqlalchemy.sql.selectable.Alias at 0x7ff5d311b080; alias_6> being replaced by Column('name', TEXT(), table=<alias_6>), which has the same key.  Consider use_labels for select() statements.
  self[column.key] = column

Out[13]: 19524825


```

**Comment by cpcloud** [comment]
@FRidh, Can you provide a link to this dataset, or a small subset of it? I'd like to debug this and see what's going on. Thanks!

**Comment by mrocklin** [comment]
Hrm, that is strange.  I'll echo @cpcloud 's request for the dataset (are the RDatasets all available in sqlite form somewhre?)

I'd also be interested in seeing the generated SQL for line 11.

```python
>>> from blaze import compute
>>> print compute(join( d.event[d.event.include==True], d.operation[['name', 'datetime_nearest_close']], 'operation', 'name').count())
```

**Comment by FRidh** [comment]
Here is the generated SQL
```py

In [1]: from blaze import *

In [2]: d = Data('sqlite:///sonair.sqlite3')

In [3]: print( compute(join( d.event[d.event.include==True], d.operation[['name', 'datetime_nearest_close']], 'operation', 'name').count()) )
SELECT count(alias_1.operation) AS __count 
FROM (SELECT event.operation AS operation, event.name AS name, event.datetime_nearest_receiver AS datetime_nearest_receiver, event.aircraft AS aircraft, event.temperature_2m AS temperature_2m, event.temperature_5cm AS temperature_5cm, event.humidity AS humidity, event.windspeed AS windspeed, event.pressure AS pressure, event.include AS include, operation.datetime_nearest_close AS datetime_nearest_close 
FROM event, operation, (SELECT event.name AS name, event.operation AS operation, event.datetime_nearest_receiver AS datetime_nearest_receiver, event.aircraft AS aircraft, event.temperature_2m AS temperature_2m, event.temperature_5cm AS temperature_5cm, event.humidity AS humidity, event.windspeed AS windspeed, event.pressure AS pressure, event.include AS include 
FROM event 
WHERE event.include = 1) JOIN (SELECT operation.name, operation.datetime_nearest_close 
FROM operation) ON event.operation = operation.name) AS alias_1

```

The dataset is not publicly available. I could create a file containing a small subset. What is the recommended way to select a certain amount of rows? Should I go the `sqlite -> DataFrame.loc -> sqlite` route for this?

**Comment by FRidh** [comment]
I'm using the following versions:
```
blaze                     0.7.2                np19py34_0  
blaze-core                0.7.2                np19py34_0  
into                      0.2.2                np19py34_0 
```

**Comment by FRidh** [comment]
Here the [link](https://de.owncube.com/public.php?service=files&t=415d715e3c25234f8f67e3eed86ffc37&download) to a subset. It shows the same behaviour.

```py
In [10]: d = Data('sqlite:///subset.sqlite3')

In [11]: join( d.event[d.event.include==True], d.operation[['name', 'datetime_nearest_close']], 'operation', 'name').count()
Out[11]: 210
```

**Comment by mrocklin** [comment]
Thanks for the subset.  Looking at the SQL below gives a clue as to what's going wrong

```sql
Good version

SELECT count(alias_8.operation) AS __count
FROM (SELECT event.operation AS operation, event.name AS name, 
             event.datetime_nearest_receiver AS datetime_nearest_receiver, 
             event.aircraft AS aircraft, 
             event.temperature_2m AS temperature_2m, 
             event.temperature_5cm AS temperature_5cm, 
             event.humidity AS humidity, 
             event.windspeed AS windspeed, 
             event.pressure AS pressure, 
             event.include AS include, 
             operation.runway AS runway, 
             operation.takeoff AS takeoff, 
             operation.datetime_nearest_close AS datetime_nearest_close
       FROM event JOIN operation ON event.operation = operation.name
       WHERE event.include = 1) AS alias_8

Bad version

SELECT count(alias_9.operation) AS __count
FROM (SELECT event.operation AS operation, 
             event.name AS name, 
             event.datetime_nearest_receiver AS datetime_nearest_receiver, 
             event.aircraft AS aircraft, 
             event.temperature_2m AS temperature_2m, 
             event.temperature_5cm AS temperature_5cm, 
             event.humidity AS humidity, 
             event.windspeed AS windspeed, 
             event.pressure AS pressure, 
             event.include AS include, 
             operation.datetime_nearest_close AS datetime_nearest_close
      FROM event, operation, (SELECT event.name AS name, 
                                     event.operation AS operation, 
                                     event.datetime_nearest_receiver AS datetime_nearest_receiver,                                           event.aircraft AS aircraft, 
                                     event.temperature_2m AS temperature_2m, 
                                     event.temperature_5cm AS temperature_5cm, 
                                     event.humidity AS humidity, 
                                     event.windspeed AS windspeed, 
                                     event.pressure AS pressure, 
                                     event.include AS include
                                 FROM event           
                                 WHERE event.include = 1) JOIN (SELECT operation.name, 
                                                                       operation.datetime_nearest_close
                                                                 FROM operation) 
                                                            ON event.operation = operation.name) 
                                                            AS alias_9
```

It's joining on the event table twice, hence the blowup.  I'll try to track this down today.

**Comment by mrocklin** [comment]
This is somewhat solved in #979 I might play around with the solution a bit more though to see if I can clean it up.

**Comment by mrocklin** [comment]
@FRidh I believe that this is fixed in #979.  

Depending on your technology of choice you can update with the following:

    conda install -c blaze blaze
    pip install --upgrade git+https://github.com/ContinuumIO/blaze
    git pull 

Please do let me know if this doesn't resolve your issue and thank you again for reporting the bug.
